core/drivers/clk/clk-stm32mp21.c

```diff
@@ -0,0 +1,3759 @@
+	uintptr_t pllxcfgr6 = pllxcfgr1 + RCC_OFFSET_PLLXCFGR6;
+	uintptr_t pllxcfgr7 = pllxcfgr1 + RCC_OFFSET_PLLXCFGR7;
+	unsigned long dfout = 0;
+	uint32_t postdiv1 = 0;
+	uint32_t postdiv2 = 0;
+
+	postdiv1 = io_read32(pllxcfgr6) & RCC_PLLxCFGR6_POSTDIV1_MASK;
+	postdiv2 = io_read32(pllxcfgr7) & RCC_PLLxCFGR7_POSTDIV2_MASK;
+
+	if ((io_read32(pllxcfgr4) & RCC_PLLxCFGR4_BYPASS) ||
+	    !postdiv1 || !postdiv2)
+		dfout = prate;
+	else
+		dfout = clk_get_pll_fvco(cfg->pll_offset,
+					 prate) / (postdiv1 * postdiv2);
+
+	return dfout;
+}
+
+static TEE_Result clk_stm32_pll_enable(struct clk *clk)
+{
+	struct clk_stm32_pll_cfg *cfg = clk->priv;
+
+	if (stm32_rcc_has_access_by_id(RCC_RIF_PLL4_TO_8)) {
+		if (stm32_gate_rdy_enable(cfg->gate_id)) {
+			EMSG("%s timeout", clk_get_name(clk));
+			return TEE_ERROR_TIMEOUT;
+		}
+	}
+
+	return TEE_SUCCESS;
+}
+
+static void clk_stm32_pll_disable(struct clk *clk)
+{
+	struct clk_stm32_pll_cfg *cfg = clk->priv;
+
+	if (stm32_rcc_has_access_by_id(RCC_RIF_PLL4_TO_8))
+		if (stm32_gate_rdy_disable(cfg->gate_id) != 0U)
+			EMSG("%s timeout", clk_get_name(clk));
+}
+
+static const struct clk_ops clk_stm32_pll_ops = {
+	.get_parent = clk_stm32_pll_get_parent,
+	.get_rate = clk_stm32_pll_get_rate,
+	.enable = clk_stm32_pll_enable,
+	.disable = clk_stm32_pll_disable,
+};
+
+struct clk_stm32_flexgen_cfg {
+	int flex_id;
+};
+
+static size_t clk_stm32_flexgen_get_parent(struct clk *clk)
+{
+	struct clk_stm32_flexgen_cfg *cfg = clk->priv;
+	uintptr_t rcc_base = clk_stm32_get_rcc_base();
+	uint32_t address = 0;
+
+	address = rcc_base + RCC_XBAR0CFGR + (cfg->flex_id * 4);
+
+	return io_read32(address) & RCC_XBAR0CFGR_XBAR0SEL_MASK;
+}
+
+static TEE_Result clk_stm32_flexgen_set_parent(struct clk *clk, size_t pidx)
+{
+	uintptr_t rcc_base = clk_stm32_get_rcc_base();
+	struct clk_stm32_flexgen_cfg *cfg = clk->priv;
+	uint16_t channel = cfg->flex_id * 4;
+
+	if (stm32_rcc_has_access_by_id(cfg->flex_id)) {
+		io_clrsetbits32(rcc_base + RCC_XBAR0CFGR + (channel),
+				RCC_XBAR0CFGR_XBAR0SEL_MASK, pidx);
+
+		if (wait_xbar_sts(channel))
+			return TEE_ERROR_GENERIC;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static unsigned long clk_stm32_flexgen_get_rate(struct clk *clk __unused,
+						unsigned long prate)
+{
+	struct clk_stm32_flexgen_cfg *cfg = clk->priv;
+	uintptr_t rcc_base = clk_stm32_get_rcc_base();
+	uint32_t prediv = 0;
+	uint32_t findiv = 0;
+	uint8_t channel = cfg->flex_id;
+	unsigned long freq = prate;
+
+	prediv = io_read32(rcc_base + RCC_PREDIV0CFGR + (0x4 * channel)) &
+		RCC_PREDIV0CFGR_PREDIV0_MASK;
+	findiv = io_read32(rcc_base + RCC_FINDIV0CFGR + (0x4 * channel)) &
+		RCC_FINDIV0CFGR_FINDIV0_MASK;
+
+	if (!freq)
+		return 0;
+
+	switch (prediv) {
+	case 0x0:
+		break;
+
+	case 0x1:
+		freq /= 2;
+		break;
+
+	case 0x3:
+		freq /= 4;
+		break;
+
+	case 0x3FF:
+		freq /= 1024;
+		break;
+
+	default:
+		EMSG("Unsupported PREDIV value (%#"PRIx32")", prediv);
+		panic();
+		break;
```

Comment:
nitpicking: by the way, remove the `break;` instruction. One may think `panic()` can return :)